File: etask-python/tests/test_preamble.py
# tools/tests/test_preamble.py
# SPDX-License-Identifier: MIT

import pytest

from etask.preamble import (
    MAGIC,
    SIZE,
    VERSION,
    PreambleError,
    SchemaMismatch,
    decode,
    discardable,
    encode,
    find,
)

_FP = 0x71DD4EB1C4E0392D

#: The exact bytes `etask/core/protocol/preamble.hpp` produces for _FP, captured
#: from a compiled probe. This is the cross-language contract: if either side
#: moves, this test fails rather than the handshake failing in the field for a
#: reason nobody can see.
_CPP_BYTES = bytes([
    0x45, 0x54, 0x53, 0x4B,   # "ETSK"
    0x01,                     # version
    0x00,                     # reserved
    0x71, 0xDD, 0x4E, 0xB1, 0xC4, 0xE0, 0x39, 0x2D,   # fingerprint, big-endian
])


# ------------------------------------------------------- the wire contract

def test_matches_the_cpp_encoding_byte_for_byte():
    assert encode(_FP) == _CPP_BYTES


def test_is_fourteen_bytes():
    assert SIZE == 14
    assert len(encode(_FP)) == 14


def test_layout_is_where_the_header_says_it_is():
    raw = encode(_FP)
    assert raw[0:4] == MAGIC
    assert raw[4] == VERSION
    assert raw[5] == 0
    assert int.from_bytes(raw[6:14], "big") == _FP


def test_fingerprint_is_big_endian():
    # A hex dump must read the same as the digest, on either peer.
    raw = encode(0x0123456789ABCDEF)
    assert raw[6:14] == bytes([0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF])


@pytest.mark.parametrize("value", [0, 1, 0xFF, 0xFFFFFFFF, 0xFFFFFFFFFFFFFFFF])
def test_round_trips_edge_values(value):
    error, peer = decode(encode(value), value)
    assert error is PreambleError.NONE
    assert peer == value


# ----------------------------------------------------------------- decoding

def test_accepts_a_matching_peer():
    error, peer = decode(encode(_FP), _FP)
    assert error is PreambleError.NONE
    assert peer == _FP


def test_rejects_a_different_schema_and_reports_it():
    error, peer = decode(encode(0xDEADBEEFCAFEBABE), _FP)
    assert error is PreambleError.FINGERPRINT_MISMATCH
    # The peer's value is the "actual" half of the log line.
    assert peer == 0xDEADBEEFCAFEBABE


def test_rejects_a_foreign_protocol():
    error, peer = decode(b"HTTP/1.1 200 " + bytes(8), _FP)
    assert error is PreambleError.BAD_MAGIC
    assert peer is None


def test_rejects_an_unknown_preamble_version():
    raw = bytearray(encode(_FP))
    raw[4] = 0xFF
    error, peer = decode(bytes(raw), _FP)
    assert error is PreambleError.BAD_VERSION
    assert peer is None


def test_does_not_report_a_fingerprint_it_could_not_trust():
    # Eight bytes read out of a non-preamble are a random number that reads like
    # a real schema id; reporting one sends an operator hunting for a schema
    # that never existed.
    for raw in (b"XXXX" + bytes(10), bytearray(encode(_FP))[:4] + b"\xFF" + bytes(9)):
        _, peer = decode(bytes(raw), _FP)
        assert peer is None


def test_rejects_a_short_read():
    error, _ = decode(encode(_FP)[:13], _FP)
    assert error is PreambleError.BAD_MAGIC


def test_ignores_the_reserved_byte():
    # A future version may use it; refusing a non-zero value here would make
    # this build reject a peer it could otherwise talk to.
    raw = bytearray(encode(_FP))
    raw[5] = 0x7F
    error, _ = decode(bytes(raw), _FP)
    assert error is PreambleError.NONE


# ------------------------------------------------------------ stream framing

def test_finds_a_preamble_at_the_start():
    assert find(encode(_FP)) == 0


def test_finds_a_preamble_after_leading_noise():
    # A peer may be mid-stream, or may predate the handshake and open with task
    # traffic; neither may desynchronise the reader permanently.
    assert find(b"\x00\x01noise" + encode(_FP)) == 7


def test_reports_no_preamble_until_it_is_complete():
    partial = encode(_FP)[:10]
    assert find(partial) == -1
    assert find(partial + encode(_FP)[10:]) == 0


def test_reports_no_preamble_in_unrelated_bytes():
    assert find(b"nothing to see here") == -1


def test_discards_bytes_that_cannot_begin_a_preamble():
    assert discardable(b"junk" + MAGIC) == 4
    assert discardable(b"junk") == 4


def test_keeps_a_partial_magic_the_next_read_might_complete():
    # "ET" could be the start of the magic; dropping it would lose the frame.
    assert discardable(b"junkET") == 4
    assert discardable(b"ETS") == 0


# -------------------------------------------------------------- the exception

def test_mismatch_error_names_both_fingerprints():
    err = SchemaMismatch(PreambleError.FINGERPRINT_MISMATCH, _FP, 0xDEADBEEFCAFEBABE)
    text = str(err)
    assert "71dd4eb1c4e0392d" in text
    assert "deadbeefcafebabe" in text
    assert "Regenerate" in text


def test_bad_magic_error_suggests_the_likely_cause():
    text = str(SchemaMismatch(PreambleError.BAD_MAGIC, _FP, None))
    assert "different protocol" in text or "older than the handshake" in text

File: tests/protocol/test_preamble.cpp
// tests/protocol/test_preamble.cpp
// SPDX-License-Identifier: MIT
//
// The handshake preamble is the one frame whose layout may never move: it is
// what two peers exchange *before* they know whether they agree about frame
// layout at all, so it cannot itself be described by the schema.
//
// `etask-python/tests/test_preamble.py` asserts the same bytes from the other
// side, against a literal captured from this implementation. Both files have to
// be wrong in the same way for a drift to go unnoticed, which is the point of
// having both.

#include <gtest/gtest.h>

#include <cstdint>
#include <cstring>

#include <etask/core/protocol/preamble.hpp>

namespace {

    using etask::core::protocol::preamble;
    using etask::core::protocol::preamble_error;

    /// A fingerprint whose every byte differs, so a byte-order bug cannot hide
    /// behind a palindrome.
    constexpr std::uint64_t kFingerprint = 0x71DD4EB1C4E0392DULL;

    /// The exact bytes the Python half expects. Copied from
    /// `etask-python/tests/test_preamble.py`; if the two ever disagree, the
    /// handshake fails in the field for a reason nobody can see.
    constexpr unsigned char kExpected[] = {
        0x45, 0x54, 0x53, 0x4B,                            // "ETSK"
        0x01,                                              // version
        0x00,                                              // reserved
        0x71, 0xDD, 0x4E, 0xB1, 0xC4, 0xE0, 0x39, 0x2D,    // fingerprint, big-endian
    };

    /// Encodes into a fresh buffer, so no test can see another's leftovers.
    struct encoded {
        std::byte bytes[preamble::size];

        explicit encoded(std::uint64_t fingerprint) noexcept
        {
            preamble::encode(bytes, fingerprint);
        }
    };

} // namespace

// ------------------------------------------------------------ the wire contract

TEST(Preamble, IsFourteenBytes)
{
    // Fixed forever: a peer reads this many bytes before it can know anything
    // else about the sender, so the number cannot be schema-dependent.
    EXPECT_EQ(preamble::size, 14u);
    EXPECT_EQ(sizeof(kExpected), preamble::size);
}

TEST(Preamble, MatchesThePythonEncodingByteForByte)
{
    const encoded frame{kFingerprint};
    EXPECT_EQ(std::memcmp(frame.bytes, kExpected, preamble::size), 0);
}

TEST(Preamble, LayoutIsWhereTheHeaderSaysItIs)
{
    const encoded frame{kFingerprint};
    const auto* raw = reinterpret_cast<const unsigned char*>(frame.bytes);

    EXPECT_EQ(std::memcmp(raw + preamble::magic_offset, "ETSK", preamble::magic_size), 0);
    EXPECT_EQ(raw[preamble::version_offset], preamble::version);
    EXPECT_EQ(raw[preamble::reserved_offset], 0u);
    EXPECT_EQ(preamble::read_fingerprint(frame.bytes), kFingerprint);
}

TEST(Preamble, FingerprintIsBigEndian)
{
    // So that a hex dump reads the same as the digest on either peer, and a log
    // line can be compared against a manual sha256sum by eye.
    const encoded frame{0x0123456789ABCDEFULL};
    const auto* raw = reinterpret_cast<const unsigned char*>(frame.bytes);

    const unsigned char want[] = {0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF};
    EXPECT_EQ(std::memcmp(raw + preamble::fingerprint_offset, want, sizeof(want)), 0);
}

TEST(Preamble, EncodeWritesEveryByte)
{
    // Documented to need no pre-clearing, so a caller's dirty buffer must not
    // leak into the frame.
    std::byte buffer[preamble::size];
    std::memset(buffer, 0xAB, sizeof(buffer));
    preamble::encode(buffer, kFingerprint);

    EXPECT_EQ(std::memcmp(buffer, kExpected, preamble::size), 0);
}

// ------------------------------------------------------------------- decoding

TEST(Preamble, AcceptsAMatchingPeer)
{
    const encoded frame{kFingerprint};
    std::uint64_t peer = 0;

    EXPECT_EQ(preamble::decode(frame.bytes, kFingerprint, &peer), preamble_error::none);
    EXPECT_EQ(peer, kFingerprint);
}

TEST(Preamble, RejectsADifferentSchemaAndReportsIt)
{
    constexpr std::uint64_t theirs = 0xDEADBEEFCAFEBABEULL;
    const encoded frame{theirs};
    std::uint64_t peer = 0;

    EXPECT_EQ(preamble::decode(frame.bytes, kFingerprint, &peer),
              preamble_error::fingerprint_mismatch);
    // The peer's value is the "actual" half of the log line, so a mismatch must
    // still hand it back - that is the whole diagnostic.
    EXPECT_EQ(peer, theirs);
}

TEST(Preamble, RejectsAForeignProtocol)
{
    std::byte frame[preamble::size]{};
    std::memcpy(frame, "HTTP/1.1 200 ", 13);

    EXPECT_EQ(preamble::decode(frame, kFingerprint), preamble_error::bad_magic);
}

TEST(Preamble, RejectsAnUnknownVersion)
{
    encoded frame{kFingerprint};
    frame.bytes[preamble::version_offset] = std::byte{0xFF};

    EXPECT_EQ(preamble::decode(frame.bytes, kFingerprint), preamble_error::bad_version);
}

TEST(Preamble, DoesNotReportAFingerprintItCouldNotTrust)
{
    // Eight bytes read out of a non-preamble are a random number that reads like
    // a real schema id; reporting one sends an operator hunting for a schema
    // that never existed. The out-param must be left exactly as it was.
    constexpr std::uint64_t untouched = 0x1111111111111111ULL;

    {
        std::byte frame[preamble::size]{};
        std::memcpy(frame, "XXXX", 4);
        std::uint64_t peer = untouched;
        EXPECT_EQ(preamble::decode(frame, kFingerprint, &peer), preamble_error::bad_magic);
        EXPECT_EQ(peer, untouched);
    }
    {
        encoded frame{kFingerprint};
        frame.bytes[preamble::version_offset] = std::byte{0xFF};
        std::uint64_t peer = untouched;
        EXPECT_EQ(preamble::decode(frame.bytes, kFingerprint, &peer),
                  preamble_error::bad_version);
        EXPECT_EQ(peer, untouched);
    }
}

TEST(Preamble, IgnoresTheReservedByte)
{
    // A future version may use it; refusing a non-zero value here would make
    // this build reject a peer it could otherwise talk to.
    encoded frame{kFingerprint};
    frame.bytes[preamble::reserved_offset] = std::byte{0x7F};

    EXPECT_EQ(preamble::decode(frame.bytes, kFingerprint), preamble_error::none);
}

TEST(Preamble, AcceptsANullOutParameter)
{
    // Documented as optional for a caller that only wants the verdict.
    const encoded frame{kFingerprint};
    EXPECT_EQ(preamble::decode(frame.bytes, kFingerprint, nullptr), preamble_error::none);
}

// ------------------------------------------------------------------ round trip

TEST(Preamble, RoundTripsEdgeValues)
{
    for (const std::uint64_t value : {std::uint64_t{0},
                                      std::uint64_t{1},
                                      std::uint64_t{0xFF},
                                      std::uint64_t{0xFFFFFFFF},
                                      ~std::uint64_t{0}}) {
        const encoded frame{value};
        std::uint64_t peer = 0;

        EXPECT_EQ(preamble::decode(frame.bytes, value, &peer), preamble_error::none)
            << "value " << value;
        EXPECT_EQ(peer, value) << "value " << value;
    }
}
